zebra/zebra_mpls.c

```diff
@@ -197,10 +200,29 @@ static int lsp_install(struct zebra_vrf *zvrf, mpls_label_t label,
 	 */
 	for (nexthop = re->nhe->nhg.nexthop;
 	     nexthop; nexthop = nexthop->next) {
+		nexthop_via_recursive = nexthop;
 		/* Skip inactive and recursive entries. */
 		if (!CHECK_FLAG(nexthop->flags, NEXTHOP_FLAG_ACTIVE))
 			continue;
-		if (CHECK_FLAG(nexthop->flags, NEXTHOP_FLAG_RECURSIVE))
+		if (CHECK_FLAG(nexthop->flags, NEXTHOP_FLAG_RECURSIVE) &&
+		    zvrf->zebra_mpls_fec_recursive_seamless) {
+			if (nexthop->type == NEXTHOP_TYPE_IPV4 ||
+			    nexthop->type == NEXTHOP_TYPE_IPV4_IFINDEX) {
+				p.family = AF_INET;
+				memcpy(&p.u.prefix4, &nexthop->gate.ipv4,
+				       sizeof(p.u.prefix4));
+				p.prefixlen = IPV4_MAX_BITLEN;
+			} else if (nexthop->type == NEXTHOP_TYPE_IPV6 ||
+				   nexthop->type == NEXTHOP_TYPE_IPV6_IFINDEX) {
+				p.family = AF_INET6;
+				memcpy(&p.u.prefix6, &nexthop->gate.ipv6,
+				       sizeof(p.u.prefix6));
+				p.prefixlen = IPV6_MAX_BITLEN;
+			}
+
+			re_recursive = rib_lookup(&p, zvrf->vrf->vrf_id);
+			nexthop_via_recursive = re_recursive->nhe->nhg.nexthop;
```

Comment:
what if the lookup doesn't succeed?
and again, we have to have a comment describing what this branch is trying to do.

Reply:
Indeed, the code should also cope with this situation.